### PR TITLE
Add hyperlink video timestamps in description

### LINKED ISSUE #1312: Automatically hyperlink video timecodes in description and comment text
This is a request for a new feature, where common video timecode formats (like 3:96, 1:23:40) will be identified automatically and hyperlinked to a timecode within the video on the current page.

This shorthand may come naturally to people, and other platforms support this sort of hyperlinking. It's especially useful for commenting on a specific moment in a video (lacking a specific way to do this?) or for writing descriptions that include a written table-of-contents.

For example, this is a long (12 hour!) video that includes a description I originally wrote for YouTube. On that platform, clicking the timecode will seek the player to that location. Not perfect, but it's nice to have a freeform solution for use cases like this that the software doesn't directly support yet.

https://diode.zone/videos/watch/2f1544f6-3a3b-4cd0-bb72-c57d2e2d1e9b

### PR DESCRIPTION
**Pull request by Lesterpig**
Fix #1312 (duplicates: #1728 and #2007)
The modification is also applied to comments and video editing.

### CODE REVIEW

**Comment by alcalyn** [comment]
Works well ! Although for now, it is reloading the full page when clicking on a timecode

Capture d’écran - 2019-10-07 à 18 34 41

**Comment by Lesterpig** [comment]
Yep, if some has any experience with the player to avoid that refresh it would awesome :)

**Comment by alcalyn** [comment]
I can make it works by adding in `client/src/app/videos/+video-watch/video-watch.component.ts`:

``` typescript
  bindTimecodeLinks () {
    const timecodeLinks = this.elementRef.nativeElement.querySelectorAll('.timecode')

    timecodeLinks.forEach((timecodeLink: HTMLElement) => {
      this.bindTimecodeLink(timecodeLink)
    })
  }

  bindTimecodeLink (timecodeLink: HTMLElement) {
    timecodeLink.addEventListener('click', (event: MouseEvent) => {
      event.preventDefault()

      const timecode: number = this.getTimecodeFromText((event.target as HTMLInputElement).innerText)

      this.player.currentTime(timecode)
      this.player.play()
    })
  }

  getTimecodeFromText (timecode: string): number {
    const tokens: number[] = timecode.split(':').map(n => parseInt(n, 10))

    const s: number = tokens.pop()
    const m: number = tokens.pop()
    const h: number = tokens.pop()

    return s + (m ? m * 60 : 0) + (h ? h * 3600 : 0)
  }
```

and a class to your timecode links:

``` diff
-      return `<a href="${url}">${str}</a>`
+      return `<a class="timecode" href="${url}">${str}</a>`
```

But I don't know when to proper bind button, the only solution I found for now is:

``` typescript
  ngAfterViewInit () {
    setTimeout(() => {
      this.bindTimecodeLinks()
    }, 1000)
  }
```

Also, I think we could create a markdownIt plugin to move the methods inside.

**Comment by Lesterpig** [comment]
Thanks for taking the time to look at this :+1: 
I'll try to apply your method to my commits during the week-end

**Comment by Chocobozzz** [comment]
Instead of using query selectors, you could also use a directive. See for example: https://stackoverflow.com/a/49006667

**Comment by kontrollanten** [comment]
Great work! I'd propose that we accept the full page reload for now, and then creates an issue for making it smoother in the future.

File: client/src/app/shared/renderer/markdown.service.ts [hunk 1]
```diff
@@ -130,7 +139,7 @@ export class MarkdownService {
   private avoidTruncatedTags (html: string) {
     return html.replace(/\*\*?([^*]+)$/, '$1')
       .replace(/<a[^>]+>([^<]+)<\/a>\s*...((<\/p>)|(<\/li>)|(<\/strong>))?$/mi, '$1...')
-      .replace(/\[[^\]]+\]\(([^\)]+)$/m, '$1')
+      .replace(/\[[^\]]+\]?\(?([^\)]+)$/, '$1')
```
Comment (Chocobozzz) [inline client/src/app/shared/renderer/markdown.service.ts]: ?

File: client/src/app/videos/+video-watch/comment/video-comment.component.ts [hunk 2]
```diff
@@ -78,16 +79,16 @@ export class VideoCommentComponent implements OnInit, OnChanges {
   }
 
   isRemovableByUser () {
-    return this.comment.account && this.isUserLoggedIn() &&
+    return this.isUserLoggedIn() &&
```
Comment (Chocobozzz) [inline client/src/app/videos/+video-watch/comment/video-comment.component.ts]: why did you remove `this.comment.account`?

File: client/src/app/videos/+video-watch/comment/video-comment.component.ts [hunk 3]
```diff
@@ -78,16 +79,16 @@ export class VideoCommentComponent implements OnInit, OnChanges {
...
       (
         this.user.account.id === this.comment.account.id ||
         this.user.hasRight(UserRight.REMOVE_ANY_VIDEO_COMMENT)
       )
   }
 
   private async init () {
-    this.sanitizedCommentHTML = await this.markdownService.textMarkdownToHTML(this.comment.text, true)
-
+    const safeHTML = await this.htmlRenderer.toSafeHtml(this.comment.text)
```
Comment (Chocobozzz) [inline client/src/app/videos/+video-watch/comment/video-comment.component.ts]: I think there is a rebase issue here no?

File: client/src/app/shared/angular/timestamp-route-transformer.directive.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      const ngxLink = document.createElement('a')
+      ngxLink.href = $event.target.getAttribute('href')
+
+      // we only care about reflective links
+      if (ngxLink.host !== window.location.host) return
+
+      const ngxLinkParams = new URLSearchParams(ngxLink.search)
+      if (ngxLinkParams.has('start')) this.timestampClicked.emit(parseInt(ngxLinkParams.get('start'), 10))
+
+      this.router.navigateByUrl(ngxLink.pathname + ngxLink.search + ngxLink.hash)
```
Comment (Chocobozzz) [inline client/src/app/shared/angular/timestamp-route-transformer.directive.ts]: I'm not sure we should update the link (youtube does not seem to update it), because you could have some undesired effects: you handle URL param change + handle the link click in the watch page

File: client/src/app/videos/+video-watch/comment/video-comment.component.ts (hunk 2 above)
Comment (rigelk) [reply client/src/app/videos/+video-watch/comment/video-comment.component.ts]: Ah, that's an artifact from the messy conflict resolution I went through. Sorry about that.

File: client/src/app/shared/angular/timestamp-route-transformer.directive.ts (hunk 4 above)
Comment (rigelk) [reply client/src/app/shared/angular/timestamp-route-transformer.directive.ts]: I can remove the link update altogether.

Comment (Chocobozzz) [reply client/src/app/shared/angular/timestamp-route-transformer.directive.ts]: I think keeping the link is fine (in case where your open the link in another tab)

**Comment by Chocobozzz** [comment]
Thanks @rigelk & @Lesterpig
